crates/bevy_ecs/src/query/access_iter.rs

```diff
@@ -0,0 +1,515 @@
+                AccessCompatible::Compatible
+            }
+
+            // single component checks
+            (Component(Read(id)), Component(Write(id_other)))
+            | (Component(Write(id)), Component(Read(id_other)))
+            | (Component(Write(id)), Component(Write(id_other)))
+            | (
+                Resource(ResourceAccessLevel::Read(id)),
+                Resource(ResourceAccessLevel::Write(id_other)),
+            )
+            | (
+                Resource(ResourceAccessLevel::Write(id)),
+                Resource(ResourceAccessLevel::Read(id_other)),
+            )
+            | (
+                Resource(ResourceAccessLevel::Write(id)),
+                Resource(ResourceAccessLevel::Write(id_other)),
+            ) => (id != id_other).into(),
+
+            // Except* access in first parameter
+            (
+                Component(ReadAllExcept {
+                    component_id: id,
+                    index,
+                }),
+                Component(Write(id_other)),
+            )
+            | (
+                Component(WriteAllExcept {
+                    component_id: id,
+                    index,
+                }),
+                Component(Read(id_other)),
+            )
+            | (
+                Component(WriteAllExcept {
+                    component_id: id,
+                    index,
+                }),
+                Component(Write(id_other)),
+            )
+            // Except* access in second parameter
+            | (
+                Component(Write(id)),
+                Component(ReadAllExcept {
+                    component_id: id_other,
+                    index,
+                }),
+            )
+            | (
+                Component(Read(id)),
+                Component(WriteAllExcept {
+                    component_id: id_other,
+                    index,
+                }),
+            )
+            | (
+                Component(Write(id)),
+                Component(WriteAllExcept {
+                    component_id: id_other,
+                    index,
+                }),
+            ) => {
+                if id == id_other {
+                    AccessCompatible::CompatibleExcept(index)
+                } else {
+                    AccessCompatible::ConflictsExcept(index)
+                }
+            }
+
+            // WriteAll will always conflict if they have different indexes
+            (
+                Component(WriteAllExcept { index, .. }),
+                Component(WriteAllExcept {
+                    index: index_other, ..
+                }),
+            ) => (index == index_other).into(),
+        }
+    }
+}
+
+/// Error returned from [`has_conflicts`].
+#[derive(Clone, Copy, Debug, PartialEq)]
+pub enum QueryAccessError {
+    /// Component was not registered on world
+    ComponentNotRegistered,
+    /// The [`EcsAccessType`]'s conflict with each other
+    Conflict,
```

Comment:
Are we sure we don't want to panic on conflicting access?  It's based on the static type, so any given call will either always succeed or always fail.  And callers will be tempted to ignore the error payload and treat `Conflict` the same as `EntityDoesNotMatch`, which would make it harder to *notice* that the type has a conflict.

Reply:
yeah, I guess it's a safety issue like with normal systems. Hopefully with enough const rust features we'll be able to move a lot of these checks to compile time. We did get const type id's recently.

With this change I am tempted to change these apis back to being options. As you said above, it's not easy to hit the component not registered error.